src/metaopt/worker/experiment.py

```diff
@@ -0,0 +1,354 @@
+                if not attrname.startswith('_'):
+                    setattr(self, attrname, config[attrname])
+            self._id = config['_id']
+
+    def reserve_trial(self, score_handle=None):
+        """Find *new* trials that exist currently in database and select one of
+        them based on the highest score return from `score_handle` callable.
+
+        :param score_handle: A way to decide which trial out of the *new* ones to
+           to pick as *reserved*, defaults to a random choice.
+        :type score_handle: callable
+
+        :return: selected `Trial` object, None if could not find any.
+        """
+        if score_handle is not None and not callable(score_handle):
+            raise ValueError("Argument `score_handle` must be callable with a `Trial`.")
+
+        query = dict(
+            exp_name=self.name,
+            user=self.metadata['user'],
+            status='new'
+            )
+        new_trials = Trial.build(self._db.read('trials', query))
+
+        if not new_trials:
+            return None
+
+        if score_handle is None:
+            selected_trial = random.sample(new_trials, 1)[0]
+        else:
+            raise NotImplementedError("scoring will be supported in the next iteration.")
+
+        selected_trial.status = 'reserved'
+
+        self._db.write('trials', dict(selected_trial), query={'_id': selected_trial.id})
+
+        return selected_trial
+
+    def push_completed_trial(self, trial):
+        """Inform database about an evaluated `trial` with results.
+
+        :param trial: Corresponds to a successful evaluation of a particular run.
+        :type trial: `Trial`
+
+        .. note:: Change status from *reserved* to *completed*.
+        """
+        trial.status = 'completed'
+        self._db.write('trials', dict(trial), query={'_id': trial.id})
+
+    def register_trials(self, trials):
+        """Inform database about *new* suggested trial with specific parameter
+        values. Each of them correspond to a different possible run.
+
+        :type trials: list of `Trial`
+        """
+        for trial in trials:
+            trial.status = 'new'
+        self._db.write('trials', list(map(dict, trials)))
+
+    @property
+    def is_done(self):
+        """Count how many trials have been completed and compare with `max_trials`.
+
+        .. note:: To be used as a terminating condition in a ``Worker``.
+        """
+        query = dict(
+            exp_name=self.name,
+            user=self.metadata['user'],
+            status='completed'
+            )
+        num_completed_trials = len(self._db.read('trials', query, {'_id': 1}))
+        if num_completed_trials >= self.max_trials:
+            return True
+        return False
+
+    @property
+    def configuration(self):
+        """Return a copy of an `Experiment` configuration as a dictionary."""
+        if self._init_done:
+            return self._db.read('experiments', {'_id': self._id})[0]
```

Comment:
Well not exactly. Because, for example from the documentation of `algorithms` and `refers` as I have it in mind currently:
```
    refers : dict or list of `Experiment` objects, after initialization is done.
       A dictionary pointing to a past `Experiment` name, ``refers[name]``, whose
       trials we want to add in the history of completed trials we want to re-use.
```
and
```
    algorithms : dict of dicts or list of `Algorithm` objects, after initialization is done.
       Complete specification of the optimization and dynamical procedures taking
       place in this `Experiment`.
````

Do you agree with this perspective? Experiment is completely delegated with the responsibility of setting-up a working setting with this.

Reply:
I'm not sure I understand. After initialization, experiment.refers will point to _actual_ experiment objects and _actual_ algorithms?  And before initialization they will be mere strings or whatever is used to defined them in config? I can hardly see how self._db.read can return the priors.

Reply:
Database will not have the objects inside it, but the configuration as dictionaries.

Reply:
Let me push some stuff and I will refer to what I mean.